addons/product/models/product.py

```diff
@@ -176,6 +176,15 @@ def _set_template_field(self, template_field, variant_field):
             else:
                 record[variant_field] = record[template_field]
 
+    @api.depends("create_date", "write_date", "product_tmpl_id.create_date", "product_tmpl_id.write_date")
+    def _compute_concurrency_field(self):
+        # Intentionally not calling super() to involve all fields explicitly
+        for record in self:
+            record[self.CONCURRENCY_CHECK_FIELD] = max(filter(None, (
+                record.product_tmpl_id.write_date or record.product_tmpl_id.create_date,
+                record.write_date or record.create_date,
+            )))
```

Comment:
Can `write_date` actually be falsy? aren't those `or` condition never reached?
If so, that should do the same:
```suggestion
            record[self.CONCURRENCY_CHECK_FIELD] = max(
                record.product_tmpl_id.write_date, record.write_date
            )
```

It not, if those value can indeed be falsy and I am wrong, then the current code could just have both 4 values falsy and should fallback on `now()`, otherwise `max()` will crash since it only got an empty list?
And even if only `write_date` could be falsy, then the `filter()` would be useless, isn't it?

Reply:
Indeed, I cannot find a normal way for `write_date` to be missing on the involved models (product & product_template).
It could be set to null through explicit DB sabotage though, but I am not sure such a case must be handled.
(I.e. like when we encountered null translations)

Reply:
Given that people actually do run queries on instances with unexpected side effects, I kept the version that involves all fields (similarly to what the root model's `_compute_concurrency_field` does to be safe) and also included a `now()` on one of the two branches so there is always a value for the `max`.

Reply:
Ok, should be good now then.
I think it makes the code way more complicated that what it could be if we do not consider those field to be NULL but I am not too sure how is the framework actually handling those generally (eg, expecting it to be null sometimes due to people queries).
Better safe than sorry I guess.